### ISSUE TITLE
Raft election timeout infinite loops (2.2.5 & 2.2.6)

### ISSUE THREAD

**Issue by ntquyen** [rethinkdb/rethinkdb#6038]
I have some rethinkdb clusters (3 instances/cluster) frequently got into Raft election timeout infinite loops.

Environment: CoreOS 899.15.0, Docker 1.9.1. Each cluster is containing around 60 to more than 100 tables

The clusters are running in a quite unstable network. Before it occurred, network partition might happen a few times, causing heartbeat timeouts. But while the loops was happening, there was no heartbeat timeout, the instances still connected to each other. I had to restart each instance to stop the loop. But for some clusters, restart didn't resolve. In those cases, I had to stop all 3 instances and start one by one.

**Comment by danielmewes** [comment]
@ntquyen Were all instances fully connected while it was looping, or is it possible that one of the instances was not reconnecting properly (or was down)?

If there is no connected majority for the table, it is normal that the nodes try to perform elections over and over again which will fail each time.

We should be able to determine which servers were connected from the logs. Could you email me the RethinkDB logs from all servers to daniel@rethinkdb.com ?

**Comment by ntquyen** [comment]
@danielmewes I've just sent you the log_file for the day 2018-08-07. 

I'm sure that all instances were fully connected by checking the table `server_status` in all of them. And for some cluster, I restarted the whole cluster one by one, each instance was well connected with other two

**Comment by danielmewes** [comment]
Thanks @ntquyen , got your email. We'll see if we can find anything.

What is the hardware of the servers (disk drives)?

You say that the network is quite unstable. Is it possible that it had periods of high latency or high package loss / low throughput?

We were wondering if it could have been just fast enough not to trigger another heartbeat timeout (so the servers would on paper stay connected), but too slow to allow the Raft election process to go through in time.
That wouldn't explain why it started working again after you restarted all servers, but might be something worth looking into.

**Comment by ntquyen** [comment]
We are using SSD. We are facing an issue that our router just got restarted from time to time, disconnected for ~1min. Heartbeat timeout occurreds at that moment. But when the router is back,  the network is very fast, ping time mostly < 1ms

**Comment by danielmewes** [comment]
Ok that shouldn't be an issue then. I suspect that the number of tables might be related to this since our Raft stress tests have all been using small databases.
We'll see if we can reproduce this locally by creating more tables,

**Comment by ntquyen** [comment]
@danielmewes Yes, we've been having a big performance issue when having a lot of tables (> 150 tables) like I described in this issue https://github.com/rethinkdb/rethinkdb/issues/5972. It would be great if you make a quick test on this :)

**Comment by rsantosdev** [comment]
I will add my 2 cents. We have a 3 node cluster running on AWS and I can see lots of Raft election errors. Not sure if our network is unstable, but on docker cloud logs I can see some `heartbeat timeouts`. I am sending some screenshots of the services as long some data to your email @danielmewes

**Comment by rsantosdev** [comment]
I am running rethinkdb:2.3 docker image

**Comment by danielmewes** [comment]
@rsantosdev Looking at the logs you sent me, I think yours might be a different issue. As long as there are heartbeat timeouts and disconnects, seeing Raft election timeouts is normal and expected. This issue is about a case where the election timeouts keep happening even after a period of full connectivity. Is that what you are seeing too?
I'd recommend figuring out why the heartbeat timeouts are happening first. Those can be caused by network connectivity issues, or (maybe more likely) by performance issues on the servers. For example swapping on a server can often cause heartbeat timeouts.

**Comment by ntquyen** [comment]
@danielmewes Yesterday we faced the raft election timeout loops in almost all of our clusters, especially ones with many tables (~100 tables) and large dataset (20GB per instance). It happened after a network failure. The network recovered after 1 minute, but the raft election timeouts remained.
For more info, we are running 10 clusters of 3 instances, running in 3 machines(2 VMs with 8 dedicated CPUs, 1 bare metal), the total dataset is ~160GB per machine.

I would like to ask more about good/bad of our data structure design. We are splitting the data into many tables, one for each client. The main goal is to isolate the client data. The benefits of this design is faster in write (because of small index tree) and reduce the impact when a table is reconfiguring or down. However, that's what we reason base on our understanding of database design. We aren't sure which is better, splitting or not splitting in case of rethinkdb cluster. Can give us some advises?

**Comment by danielmewes** [comment]
@ntquyen We're investigating this right now, together with the data corruption issue that you encountered.
We can reproduce the issue, and probably have a fix for it (but there are still some open details). We'll be delaying RethinkDB 2.3.5 until the fix for this gets in, and then release the update.

Sorry you had to run into this.

Your point on impact in case a table goes down is valid. Performance-wise, I wouldn't expect much difference though between splitting the data up and having it all in one table. Indexes are tree structures, and operations on them have logarithmic costs in the number of documents. In practice that means that indexes remain very efficient even as the data grows, and there isn't a big performance difference between a small and a large index.

**Comment by ntquyen** [comment]
@danielmewes Thanks for you quick response! Can you give me more hint about how you reproduce it? What is actually the cause?

**Comment by danielmewes** [comment]
@ntquyen The precise course and minimal necessary steps for reproducing this are still unclear to me.

The way I can reproduce it is by spinning up a cluster of three servers, each using the `--direct-io` option and storing the data on rotational disks (both together slow down data access). Then I create 200 tables, and shard and replicate all of them to 3 shards * 3 replicas each. After shutting down the servers and starting them up again, they often (but not always) get stuck and tables never become ready.

It looks like increasing the Raft election timeout for repeated elections avoids this issue, and the elections eventually start converging on a leader (the only way for changing the timeout currently is to compile RethinkDB from source).

**Comment by ntquyen** [comment]
@danielmewes Thanks for your response, I'll keep waiting for your investigation!

I have some more specific question on how raft is implement in rethinkdb. In which case exactly the raft election of a table got timeout? 

In raft explaination, election timeout is the time a follower wait before become an candidate. This wait starts right after a heartbeat timeout occurs. And also when 2 instances receive the same votes at the same time at the same term, the leader cannot be determined, the wait happens then, too. The latter could cause the timeout loops, but this case could only happen when the number of replicas is even, right? If what I understand is correct, there is probably something else that I miss in rethinkdb raft...

**Comment by danielmewes** [comment]
@ntquyen It is indeed the third timeout in which it loops. Evenly split votes are not the only case where this triggers. The other case (which I think we are seeing here) is if simply not enough other nodes reply to the voting request on time. Issuing a vote is a comparably expensive operation, since it must be persisted in the Raft log, and therefore requires disk access. So this could potentially cause nodes to take more than the 1-2 second election timeout that RethinkDB currently uses.

Whenever a candidate doesn't get enough votes within the election timeout, it will start another term and loop.

This is also the timeout where the backoff that we are testing comes into play.

**Comment by danielmewes** [comment]
This is likely fixed in v2.3.x via 8ab14bd and next via 2398ccb, and will ship with RethinkDB 2.3.5.

If the issue shows up again after that upgrade, please let us know.

**Comment by danielmewes** [comment]
@ntquyen We've released RethinkDB 2.3.5 today, which includes the new Raft election timeout handling. In case you still see the issue after this update, please let us know.

**Comment by WvanLelyveld** [comment]
I'm now getting this issue. One of my 3 servers got disconnected, and when it is reconnected it connects fine to the other servers, but it starting new Raft elections for 951 primaries. The first ones go fine, but I think because there are that many, the rest go into a loop where new elections are started for all tables. None of them seem to actually finish anymore, so the loop really continues.
This is how the loop looks:

```
2017-06-29T09:05:05.438578519 3257.325243s info: Table 0b0def22-9307-4038-b13f-4bf9c96349a3: Raft election timed out. Starting a new election for term 280.
2017-06-29T09:05:05.438629627 3257.325293s info: Table 0f5b3fbd-ad8b-4fe3-b293-2e995637ef13: Raft election timed out. Starting a new election for term 280.
2017-06-29T09:05:05.438662814 3257.325327s info: Table b6b10a8e-fd51-4b6f-9c02-1828b6f8863d: Raft election timed out. Starting a new election for term 287.
2017-06-29T09:05:05.797487686 3257.684152s info: Table 19648a60-704b-41b8-a7af-03c3df6bcd1d: Raft election timed out. Starting a new election for term 285.
2017-06-29T09:05:06.102964601 3257.989629s info: Table 47f31506-5934-4220-8110-cdea74afdbd6: Raft election timed out. Starting a new election for term 181.
2017-06-29T09:05:06.545923010 3258.432587s info: Table 1a8ab551-2965-41fb-852b-88246d0e5524: Raft election timed out. Starting a new election for term 292.
```

This is the distribution of tables:
```
rdb03_tilaa_cloud	44 primaries, 442 secondaries	rdb03.tilaa.cloud, up for 21 hours
rdb02_cloud_tilaa_com	951 primaries, 234 secondaries	rdb02.cloud.tilaa.com, up for an hour
rdb01_cloud_tilaa_com	1032 primaries, 163 secondaries	rdb01.cloud.tilaa.com, up for 3 months
```

It would be cool if somehow there would only be a maximum number of elections at the same time to prevent this. Does this make sense? If you have a quick way to get out of this loop that would be nice too :)
